abci/proposals/validate.go

```diff
@@ -58,3 +60,116 @@ func (h *ProposalHandler) ValidateExtendedCommitInfo(
+// marshalled and contain valid prices. This function contains extra validation to be run in
+// ProcessProposal.
+// CONTACT: this function assumes that the RequestProcessProposal is not nil and has already been
+// checked by the caller.
+func (h *ProposalHandler) ValidateExtendedCommitInfoProcess(
+	ctx sdk.Context,
+	req *cometabci.RequestProcessProposal,
+	extendedCommitInfo cometabci.ExtendedCommitInfo,
+) error {
+	if err := h.validateVoteExtensionsFn(ctx, extendedCommitInfo); err != nil {
+		h.logger.Error(
+			"failed to validate vote extensions; vote extensions may not comprise a supermajority",
+			"height", req.Height,
+			"err", err,
+		)
+
+		return err
+	}
+
+	// checks for consistency between extendedCommitInfo and proposedLastCommit
+	if extendedCommitInfo.Round != req.ProposedLastCommit.Round {
+		h.logger.Error(
+			"mismatched round in encoded extended commit info and proposed last commit",
+			"height", req.Height,
+			"extended commit round", extendedCommitInfo.Round,
+			"proposed last commit round", req.ProposedLastCommit.Round,
+		)
+
+		return fmt.Errorf("mismatched round in encoded extended commit info and proposed last commit: %d != %d",
+			extendedCommitInfo.Round,
+			req.ProposedLastCommit.Round,
+		)
+	}
+
+	if len(extendedCommitInfo.Votes) != len(req.ProposedLastCommit.Votes) {
+		h.logger.Error(
+			"mismatched length in encoded extended commit info and proposed last commit",
+			"height", req.Height,
+			"extended commit length", len(extendedCommitInfo.Votes),
+			"proposed last commit length", len(req.ProposedLastCommit.Votes),
+		)
+
+		return fmt.Errorf("mismatched length in encoded extended commit info and proposed last commit")
+	}
+
+	requestCommits := make(map[string]cometabci.VoteInfo, len(extendedCommitInfo.Votes))
+	for _, vote := range req.ProposedLastCommit.Votes {
+		requestCommits[string(vote.Validator.Address)] = vote
+	}
+
+	// Validate all oracle vote extensions.  And cross-reference them with the ProposedLastCommit
+	for _, vote := range extendedCommitInfo.Votes {
+		var address sdk.ConsAddress = vote.Validator.Address
+		reqVote, found := requestCommits[string(address)]
+		if !found {
+			h.logger.Error(
+				"no vote for validator in extended commit vote found in proposed last commit",
+				"height", req.Height,
+				"validator", string(address),
+			)
+
+			return fmt.Errorf("no vote for validator in extended commit vote found in proposed last commit")
+		}
+
+		if reqVote.Validator.Power != vote.Validator.Power {
+			h.logger.Error(
+				"mismatched validator power between extended commit vote and last proposed commit",
+				"height", req.Height,
+				"validator", string(address),
+				"extended vote power", vote.Validator.Power,
+				"last proposed vote power", reqVote.Validator.Power,
+			)
+
+			return fmt.Errorf("mismatched validator power between extended commit vote and last proposed commit")
+		}
+
+		if reqVote.BlockIdFlag != vote.BlockIdFlag {
```

Comment:
I looked into this again, this actually is correct, the proposedBlock.LastCommit has each vote's BlockFlagID

Reply:
I'd prefer that we just add this check to the `ValidateVoteExtensions` from the SDK (and move the upstream impl into our codebase)